File: 31-40/33. Search in Rotated Sorted Array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
    let len = nums.length;
    let left = 0,
        right = len - 1;

    while (left < right) {
        let mid = left + Math.floor((right - left + 1) / 2);
        if(nums[mid] == target) return mid;
        // left side is sorted
        if (nums[left] < nums[mid]) {
            if (nums[left] <= target && target <= nums[mid]) {
                right = mid - 1;
            } else {
                left = mid;
            }
        }
        // right side is sorted
        else {
            // target can equal to nums value
            if (nums[mid] <= target && target <= nums[right]) {
                left = mid;
            } else {
                right = mid - 1;
            }
        }
    }

    return nums[left] == target ? left : -1;
};

File: 31-40/33. Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                return mid

            # left portion
            # equal is for one element case
            if nums[mid] >= nums[l]:
                # out of range [l, mid]
                if target < nums[l] or target > nums[mid]:
                    l = mid + 1
                else:
                    r = mid - 1

            # right portion
            else:
                # out of range [mid, r]
                if target < nums[mid] or target > nums[r]:
                    r = mid - 1
                else:
                    l = mid + 1

        return -1
